packages/cardano-services/src/DbSyncProvider.ts

```diff
@@ -3,19 +3,113 @@ import { Pool } from 'pg';
+>(
+  BaseClass?: T
+) => {
+  abstract class Mixin extends (BaseClass || Object) implements Provider {
+    public db: Pool;
+
+    constructor(...args: AnyArgs) {
+      const [db, ...baseArgs] = [...args];
+
+      super(...baseArgs);
+
+      this.db = db;
+    }
+
+    public async healthCheck(): Promise<HealthCheckResponse> {
+      const result = await this.db.query(HEALTH_CHECK_QUERY);
+      return { ok: !!result.rowCount };
+    }
   }
 
-  public async healthCheck(): Promise<HealthCheckResponse> {
-    const result = await this.db.query(HEALTH_CHECK_QUERY);
-    return { ok: !!result.rowCount };
+  type BaseArgs = T extends abstract new (...baseArgs: infer A) => {}
+    ? A
+    : T extends new (...baseArgs: infer A) => {}
+    ? A
+    : never;
+  type BaseInstance = T extends abstract new (...baseArgs: AnyArgs) => infer I
+    ? I
+    : T extends new (...baseArgs: AnyArgs) => infer I
+    ? I
+    : never;
+  type ReturnedType = BaseInstance & { db: Pool; healthCheck: () => Promise<HealthCheckResponse> };
+
+  return Mixin as unknown as (T extends new (...baseArgs: AnyArgs) => {}
+    ? new (db: Pool, ...args: BaseArgs) => ReturnedType
+    : abstract new (db: Pool, ...args: BaseArgs) => ReturnedType) & {
+    prototype: { healthCheck: () => Promise<HealthCheckResponse> };
+  };
+};
+
+export type DbSyncProvider<T> = T extends new (...args: AnyArgs) => infer O ? O : never;
+
+// TODO: delete following
+// Some inline use case used while development
+async () => {
+  class TestClass {
+    constructor(test: number, check: string) {
+      // eslint-disable-next-line no-console
+      console.log(check + test);
+    }
+
+    test(test: string) {
+      return test;
+    }
   }
 
-  public async start(): Promise<void> {
-    return Promise.resolve();
+  abstract class AbstractClass {
+    constructor(test: string, check: number) {
+      // eslint-disable-next-line no-console
+      console.log(test + check);
+    }
+
+    abstract test(test: string): string;
   }
-}
+
+  const db = null as unknown as Pool;
+
+  const TestProvider1 = DbSyncProvider();
+  type TestProvider1 = DbSyncProvider<typeof TestProvider1>;
+
+  const testProvider1 = new TestProvider1(db);
+  const test1 = await testProvider1.healthCheck();
+
+  const TestProvider2 = DbSyncProvider(TestClass);
+  type TestProvider2 = DbSyncProvider<typeof TestProvider2>;
+
+  const testProvider2 = new TestProvider2(db, 23, 'test');
+  const test2 = [await testProvider2.healthCheck(), testProvider2.test('test')] as const;
+
+  class TestProvider3 extends DbSyncProvider(TestClass) {
+    check(check: number) {
+      return check;
+    }
+  }
+
+  const testProvider3 = new TestProvider3(db, 23, 'test');
+  const test3 = [await testProvider3.healthCheck(), testProvider3.test('test'), testProvider3.check(23)] as const;
+
+  const BaseTestProvider4 = DbSyncProvider(AbstractClass);
+
+  // Uncomment this to check
+  // const baseTestProvider4 = new BaseTestProvider4();
```

Comment:
Why is this commented out?

Reply:
It is a sort of preview test about the constructor produced by the mixin: uncommenting it, it will rise a compile error about missing parameters

Reply:
@mkazlauskas , 0c1b5d5bc should resolve this.

Reply:
sorry @mirceahasegan , 2f2c8e7d4 should really solve